Userland/Libraries/LibGL/SoftwareGLContext.cpp

```diff
@@ -628,6 +628,37 @@ void SoftwareGLContext::gl_disable(GLenum capability)
     }
 }
 
+void SoftwareGLContext::gl_gen_textures(GLsizei n, GLuint *textures)
+{
+    size_t tex_array_index = 0;
+
+    while(n-- > 0)
+    {
+        // We have an already allocated free texture name. Pop it into the array
+        if (!m_free_texture_names.is_empty()) {
+            textures[tex_array_index++] = m_free_texture_names.dequeue();
+            continue;
+        }
+
+        // We're out of free previously allocated names. Let's allocate a new contiguous amount from the
+        // last known texture name
+        // 'n' now is the remaining number of names to allocate
+        for (auto name = m_last_texture_id; name < m_last_texture_id + n; name++) {
+            textures[tex_array_index++] = name;
+        }
+
+        m_last_texture_id += n;
+    }
+}
```

Comment:
This code is a bit hard to read, I think it would be better to do it as:

```suggestion
void SoftwareGLContext::gl_gen_textures(GLsizei n, GLuint *textures) {
    for (size_t i = 0; i < n; ++i) {
        if (!m_free_textures_names.is_empty()) {
            // We have an already allocated free texture name. Pop it into the array
            textures[i] = m_free_texture_names.dequeue();
        } else {
            // We're out of allocated texture names, generate a new one
            textures[i] = m_last_texture_id++;
        }
    }
}
```

Reply:
Out of curiosity, is it even necessary to reuse texture names?

Reply:
https://www.khronos.org/registry/OpenGL/specs/gl/glspec15.pdf

The 1.5 spec for `glGenTextures` states `returns n previously  unused  texture  object  names  in textures.   These  names  are marked as used, for the purposes of GenTextures only, but they acquire texture state  and  a  dimensionality  only  when  they  are  first  bound,  just  as  if  they  were unused.`

And for `glDeleteTextures` `After a texture object is deleted, it has no contents or dimensionality, and its name is again unused.  `

So I guess it's not strictly necessary to reuse texture names?

Reply:
TBH, I think it would be sufficient to just have an atomic counter, the only issue is if you have a very long running program that constantly asks for texture names.

Reply:
FWIW, I just tested creating and deleting 10 textures in a loop and always got the same 10 texture names back. Windows 10, NVIDIA GTX 1080.

Reply:
> Out of curiosity, is it even necessary to reuse texture names?

Tehcnically, no, but as @sunverwerth and @ccapitalK have said, it would be breaking the spec + creating behaviour that isn't seen on other OpenGL implementations.